Changes supporting falling actions

mugmuggy opened the pull request:
<img width="314" height="151" alt="Screenshot 2026-06-15 at 14 11 37" src="https://github.com/user-attachments/assets/55fa8037-8c23-4a61-9f51-8fd042cb27b2" />


*Fixes #476*

**Describe what the proposed change does**
- Corrects animations for Female patient (this probably doesn't work if they are a Stripped version at the time)
- Fixes the must happen to be in the constructor, otherwise any setNextAction calls will remove them
- Patients randomly choose a fall direction then that direction doesn't change, the random would have them fall one way, lay another and then stand up again in another
- Queues the actions more correctly for falling

I didn't activate the falling code in world, so these don't actually initiate anything but can be used for testing.

Doesn't as pointed out in the discussion #1846 attempt to address the issue of interrupted actions being incorrectly requeued for example, walking to room that is destroyed at the same time the fall action starts, which would requeue the interrupted walk to that room.

I had put a has_interrupted flag on the action and detected it as not to requeue, but not all interruptions are driven through setNextAction (interrupt_head in queue for example) and I think the other issue was actions like the walk which starts and does something, sets the interrupt handler, which means the action starts, rather than a todo_interrupt being detected to basically skip the action in the action start (seek_toilets_action_start for example and this is same issue with teleporting the drinks machine or jumping to/from bench when queue actions are interrupted as use_object also must start)

mugmuggy committed:
Changes supporting falling actions

lewri commented:
Iirc patients can't fall inside rooms. So think you should be fine unless females are stripping around the hospital

lewri reviewed (commented):
Maybe out of scope a little but falling's line 28 should be a colon and not a full stop (I think)

i.e. this line needs correcting
```lua
self.setMustHappen(true)
```

mugmuggy commented:
Forgot to copy that one across. The Stripped Female Patients won't get the updated on_ground animation in the afterload, so once they become normal ones, they will be restored with the old one, also if they transition from Standard to Stripped then back to Standard. Obviously not a problem on new saves.

lewri commented:
@mugmuggy May need to be it's own issue but it looks like some recent update triggering falls is causing the adviser to repeatedly saying the "These people are ill" "How would you like it" etc. when the falling is on.

lewri commented:
I think we lost the GitHub logs on this one, if you wanted to try and regenerate them.

Alberth289346 commented on CorsixTH/Lua/entities/humanoids/patient.lua:
```diff
@@ -381,7 +375,8 @@ end
 function Patient:shakeFist()
   if self.shake_fist_anim then
-    self:queueAction(ShakeFistAction(), 1)
+    self:queueAction(ShakeFistAction(), 4)
```
There is exactly one use of this function, so not much need for having it as a function.

lewri replied:
Could you provide a suggestion for this code spot? Lets see if we can get this one in a tidier place and we can merge what we have since falling is still disabled and for debugging only.

Alberth289346 replied:
``` bash
$ ag shakeFist 

CorsixTH/Lua/entities/humanoids/patient.lua
377:      self:shakeFist()
399:function Patient:shakeFist()
```

Usage is thus at line 377, maybe change it to something like
``` lua
    if math.random(1, 5) == 3 and self.shake_fist_anim then
      self:queueAction(ShakeFistAction(), 4)
      self:interruptAndRequeueAction(current, 5)
    else
      self:interruptAndRequeueAction(current, 4)
    end
```

Alberth289346 replied:
This also helps in understanding the magic 4 :)

lewri replied:
> EDIT: Hard-coding the queue position in the middle of the action queue makes it quite unusable. If you want to keep this, pass the index as a parameter.

Does your suggestion resolve this too? Or should this till be factored.

Alberth289346 replied:
Nope, didn't fix that. I would have said so if it did.

Alberth289346 commented:
Looks good to me, except for the problem spotted by Lewri. My suggestion is just further improvement of the code while you are at it.

lewri commented on CorsixTH/Lua/entities/humanoids/patient.lua:
```diff
@@ -342,20 +342,14 @@ function Patient:falling()
+    else
+      self:interruptAndRequeueAction(current, 4)
     end
     self:fallingAnnounce()
```
Ah, the problem I identified a while back sits here. This ends up spamming the announcer with every person falling.
It's out of scope so just needs noting to be fixed (or fixed here, if you wanted to).

lewri replied:
In fact, the falling Adviser messages are irrelevant to earthquakes.... They're when you click a patient I believe to knock them over.

mugmuggy commented:
> > Patients randomly choose a fall direction then that direction doesn't change, the random would have them fall one way, lay another and then stand up again in another
> 
> Unfortunately I don't think this one is true. I'm still see this even with your changes.
Specifically this is the on_ground/get_up where it could change direction whilst on the ground. They still fall in a random direction, but don't flop around on the ground.

lewri commented:
> > > Patients randomly choose a fall direction then that direction doesn't change, the random would have them fall one way, lay another and then stand up again in another
> > 
> > 
> > Unfortunately I don't think this one is true. I'm still see this even with your changes.
> > Specifically this is the on_ground/get_up where it could change direction whilst on the ground. They still fall in a random direction, but don't flop around on the ground.

Hmm. I've attached a webm here showing what's happening still.

https://user-images.githubusercontent.com/20030128/172935256-198499dc-a34e-402b-9cb6-f3bb35b00531.mp4

lewri committed:
Additional falling improvements

- Allow players to push people over when right clicking on them
- Only announce a fall if the player pushed them over
- Change falling to a debug setting, defaulted to false.
- Fixes falling:28
- Renames canPeeOrPuke to canPeePukeOrFall
- Adds falling as a debug menu item

lewri commented on CorsixTH/Lua/entities/humanoids/patient.lua:
```diff
@@ -356,27 +364,22 @@ function Patient:canPeeOrPuke(current)
+      self:interruptAndRequeueAction(current, 4)
     end
-    self:fallingAnnounce()
+    if player_init then self:fallingAnnounce(true) end
```
the `true` isn't needed. Not sure what I was doing.

Alberth289346 commented on CorsixTH/Lua/entities/humanoids/patient.lua:
```diff
@@ -356,27 +364,22 @@ function Patient:canPeeOrPuke(current)
-    self:fallingAnnounce()
+    if player_init then self:fallingAnnounce() end
     self:changeAttribute("happiness", -0.05) -- falling makes you very unhappy
   else
     return
```
Invert the condition so the return can be done immediately after deciding it's not happening.

lewri replied:
`if not self.falling_anim or not self:canPeeOrPuke(current) and not self.has_fallen == 1 then return end`?

Alberth289346 replied:
Since the function ends there as well, just delete the "else return" instead?

Alberth289346 replied:
Instead of your long condition, you can break it into 3 parts, and bail out as soon as one fails.

Alberth289346 replied:
An unsure why it needs a "pee or puke" function. Probably it was re-used without adapting the name.

lewri replied:
I think there's 3 options listed here now. Which one is the preferred one here? 😅 

> An unsure why it needs a "pee or puke" function. Probably it was re-used without adapting the name.

Very old code :) https://github.com/CorsixTH/CorsixTH/commit/df84d21f926a5868f86904b7d968bdf76d20cba6

Alberth289346 replied:
> Which one is the preferred one here?

Hmm, you're asking such difficult questions :)
I'd probably go for deleting the `else return`, as it's not doing anything useful.

> Very old code :)

Point is that it's being used for something else than pee or puke. The name of the function doesn't cover all it's purposes anymore. You can extend it with the new purpose, but maybe a more general name is better.? I pondered about a name and came up with `atFreeTile` or `atOpenTile`. Not sure these are good names though.

lewri replied:
The function checks more than tiles, it has 3 steps:
1. Is the current action interruptible?
2. Does the patient have going home status?
3. Finally check if the current tile is buildable, owned, and in a hospital

It's likely the extend name option is going to be the easiest here; can't think of a good general name either.

Alberth289346 replied:
I am fine with extending the name.

Good point, so perhaps this is not just 1 function, it's 2 or 3 functions mixed into one?  
The first 2 steps seem related to the question "is it possible and does it make sense to insert a new action in the queue?", while the 3rd step seems about an empty tile (which probably should be abstracted into an `isAtEmptyTile` function, but that is out of scope here).

I am not sure about the "inHospital" check for the current usage of this function. For pee and puke it makes sense, since a handyman needs to cleanup the mess, but falling doesn't need to be cleaned. Not a serious problem though, falling patients at the middle of some road are never seen anyway.  
If you want to fix things in that function, doing that in a separate patch is likely the better idea, imho.

Alberth289346 commented:
Not sure why it needs a new section in the config file. Can't we have another setting in the "debug" settings section? May somewhat depend on the public though.

lewri commented:
> Not sure why it needs a new section in the config file. Can't we have another setting in the "debug" settings section? May somewhat depend on the public though.

As an example, for soundfont override on fluidsynth builds we don't give any indicators the setting `config.soundfont` exists. We could do that here too. Though I'm not keen on some magic undocumented flag either.

Alberth289346 reviewed (approved):
Not sure how good it is exactly, but let's find out :)

lewri commented on CorsixTH/Lua/app.lua:
```diff
@@ -29,7 +29,7 @@ local runDebugger = corsixth.require("run_debugger")
 -- Recommended: Also replace/Update the summary comment
 
-local SAVEGAME_VERSION = 188 -- Add groups to win and lose criteria.
+local SAVEGAME_VERSION = 183 -- WIP support falling actions
```
Will refresh, I just needed a new artifact for this PR

lewri commented:
Everything should be cleared up now and ready. I also made this a debug menu item for easier toggling,

TheCycoONE commented:
LGTM too

lewri commented on CorsixTH/Lua/world.lua:
```diff
@@ -594,26 +594,19 @@ function World:tickEarthquake()
       if not patient.in_room and patient.falling_anim then
-
-        -- make the patients fall
-
-        -- jpirie: this is currently disabled. Calling this function
-        -- really screws up the action queue, sometimes the patients
-        -- end up with nil action queues, and sometimes the resumed
-        -- actions throw exceptions. Also, patients in the hospital
-        -- who have not yet found reception throw exceptions after
-        -- they visit reception. Some debugging needed here to get
-        -- this working.
-
-        patient:falling()
+        patient:falling(false)
       end
     end
```
Relocate to earthquake class before merge

lewri commented on CorsixTH/Lua/earthquake.lua:
```diff
@@ -145,7 +145,18 @@ function Earthquake:tick()
+  if not TheApp.config.debug_falling then return end
+  local hospital = self.world:getLocalPlayerHospital()
+  -- loop through the patients and allow the possibility for them to fall over
+  for _, patient in ipairs(hospital.patients) do
+    if not patient.in_room and patient.falling_anim then
+      patient:falling(false)
```
Should there be a probability roll here?

Alberth289346 replied:
Hmm, "tick" function eh?  Should be a very low probability if you want to notice any of a delay, I guess.

Quick falling is perhaps more useful for testing?

lewri replied:
Looks like it's every quake tick (which different from a regular tick), which in essence is a damage cycle completed.

Potentially its fine for now. If a patient can't fall at one part of the quake they may move and be able to later on. I think if it becomes too uniform then we can adjust later

Alberth289346 replied:
I agree :)   First make it work at all before starting to fine-tune it :)
